### PR TITLE
Feature/2fa checker

### PR DESCRIPTION
**Pull request by rochamarcelo**

### CODE REVIEW

File: config/users.php [hunk 1]
```diff
@@ -118,6 +118,7 @@
         ],
     ],
     'GoogleAuthenticator' => [
+        'checker' => 'CakeDC\Users\Auth\DefaultTwoFactorAuthenticationChecker',
```
Comment (steinkel) [inline config/users.php]: use DefaultTwoFactorAuthenticationChecker::class

File: src/Auth/TwoFactorAuthenticationCheckerFactory.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * Get the two factor authentication checker
+     *
+     * @return TwoFactorAuthenticationCheckerInterface
+     */
+    public function build()
+    {
+        $className = Configure::read('GoogleAuthenticator.checker');
+        $interfaces = class_implements($className);
+        $required = 'CakeDC\Users\Auth\TwoFactorAuthenticationCheckerInterface';
```
Comment (steinkel) [inline src/Auth/TwoFactorAuthenticationCheckerFactory.php]: use TwoFactorAuthenticationCheckerInterface::class

File: src/Controller/Component/GoogleAuthenticatorComponent.php [hunk 3]
```diff
@@ -28,6 +29,11 @@ class GoogleAuthenticatorComponent extends Component
     /** @var \RobThree\Auth\TwoFactorAuth $tfa */
     public $tfa;
 
+    /**
+     * @var \CakeDC\Users\Auth\TwoFactorAuthenticationCheckerInterface
+     */
+    private $checker;
```
Comment (steinkel) [inline src/Controller/Component/GoogleAuthenticatorComponent.php]: would it be protected?

File: src/Controller/Component/UsersAuthComponent.php [hunk 4]
```diff
@@ -55,7 +56,7 @@ public function initialize(array $config)
             $this->_loadRememberMe();
         }
 
-        if (Configure::read('Users.GoogleAuthenticator.login')) {
+        if ((new TwoFactorAuthenticationCheckerFactory())->build()->isEnabled()) {
```
Comment (steinkel) [inline src/Controller/Component/UsersAuthComponent.php]: break this line, use protected method for mocking

File: src/Controller/Traits/LoginTrait.php [hunk 5]
```diff
@@ -180,6 +180,7 @@ public function login()
                 return;
             }
             $user = $this->Auth->identify();
+            $googleAuthenticatorLogin = (new TwoFactorAuthenticationCheckerFactory())->build()->isRequired($user);
```
Comment (steinkel) [inline src/Controller/Traits/LoginTrait.php]: break like, use protected method for mocking

File: src/Controller/Traits/LoginTrait.php [hunk 6]
```diff
@@ -211,7 +212,7 @@ public function login()
      */
     public function verify()
     {
-        if (!Configure::read('Users.GoogleAuthenticator.login')) {
+        if (!(new TwoFactorAuthenticationCheckerFactory())->build()->isEnabled()) {
```
Comment (steinkel) [inline src/Controller/Traits/LoginTrait.php]: break line, use protected method for mocking

File: tests/TestCase/Controller/Component/GoogleAuthenticatorComponentTest.php [hunk 7]
```diff
@@ -41,6 +42,7 @@ class GoogleAuthenticatorComponentTest extends TestCase
     public function setUp()
     {
         parent::setUp();
+        Configure::write('Error.errorLevel', E_ALL & ~E_DEPRECATED);
```
Comment (steinkel) [inline tests/TestCase/Controller/Component/GoogleAuthenticatorComponentTest.php]: remove this line, fix specific versions in composer instead

File: src/Controller/Component/GoogleAuthenticatorComponent.php (hunk 3 above)
Comment (rochamarcelo) [reply src/Controller/Component/GoogleAuthenticatorComponent.php]: Removed unused property

**Pull request merged**